### PR TITLE
changes needed for pilot running at Summit

### PR DESCRIPTION
**Pull request by dougbenjamin**

### CODE REVIEW

File: pilot/util/harvester.py [hunk 1]
```diff
@@ -20,6 +20,11 @@
 logger = logging.getLogger(__name__)
 
 
+def dump(obj):
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: Is this function used in some external script? Can you provide the documentation field?

File: pilot/util/harvester.py [hunk 2]
```diff
@@ -126,11 +132,14 @@ def get_event_status_file(args):
...
     :return: event staus file name.
     """
+
+    logger.debug('config.Harvester.__dict__ : {0}'.format(config.Harvester.__dict__))
+
     if args.harvester_workdir != '':
         work_dir = args.harvester_workdir
     else:
         work_dir = environ['PILOT_HOME']
-    event_status_file = config.Harvester.StageOutnFile
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: Ok to rename the field, but you need to update the default.cfg file as well then

File: pilot/util/harvester.py [hunk 3]
```diff
@@ -145,11 +154,14 @@ def get_worker_attributes_file(args):
...
     """
+
+    logger.debug('config.Harvester.__dict__ : {0}'.format(config.Harvester.__dict__))
+
     if args.harvester_workdir != '':
         work_dir = args.harvester_workdir
     else:
         work_dir = environ['PILOT_HOME']
-    worker_attributes_file = config.Harvester.workerAttributesFile
+    worker_attributes_file = config.Harvester.workerattributesfile
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: same comment as above

File: pilot/util/harvester.py [hunk 4]
```diff
@@ -282,10 +310,17 @@ def publish_job_report(job, args, job_report_file="jobReport.json"):
...
-        write_json(dst_file, job_report)
+        if write_json(dst_file, job_report):
+            return True
+        else:
+            return False
 
     except IOError:
         logger.error("job report copy failed")
+        return False
+    #except:
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: Remove these lines. Flake8 has a bug that complains also about comments..

File: pilot/util/harvester.py [hunk 5]
```diff
@@ -244,19 +256,34 @@ def publish_work_report(work_report=None, worker_attributes_file="worker_attribu
...
+            if write_json(worker_attributes_file, work_report):
+                logger.info("work report published: {0}".format(work_report))
+                return True
+            else:
+                logger.info("work report publish failed: {0}".format(work_report))
+                return False
+        except IOError:
+            logger.error("job report copy failed")
+            return False
+        except:
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: Blind exception, Flake8 complains - add change to 'except Exception as e:' and then logger.error('write json file failed: %s' % e)

File: pilot/util/harvester.py [hunk 6]
```diff
@@ -244,19 +256,34 @@ def publish_work_report(work_report=None, worker_attributes_file="worker_attribu
...
+                del(work_report["outputfiles"])
+            if "inputfiles" in work_report:
+                del (work_report["inputfiles"])
+            if "xml" in work_report:
+                del (work_report["xml"])
+            if write_json(worker_attributes_file, work_report):
+                logger.info("work report published: {0}".format(work_report))
+                return True
+            else:
+                logger.info("work report publish failed: {0}".format(work_report))
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: change to logger.warning

File: pilot/control/job.py [hunk 7]
```diff
@@ -233,21 +233,38 @@ def send_state(job, args, state, xml=None, metadata=None):
...
+                log.info('Failed to write to workerAttributesFile %s' % path)
                 return False
+            # check if we are in final state then write out information for output files
+            if final:
+                # Use the job information to write Harvester event_status.dump file
+                event_status_file = get_event_status_file(args)
+                if publish_stageout_files(job, event_status_file):
+                    log.info('wrote log and output files to file %s' % event_status_file)
+                else:
+                    log.info('Warning - could not write log and output files to file %s' % event_status_file)
```
Comment (PalNilsson) [inline pilot/control/job.py]: change to logger.warning

File: pilot/control/job.py [hunk 8]
```diff
@@ -233,21 +233,38 @@ def send_state(job, args, state, xml=None, metadata=None):
...
+                    return False
+                # publish job report
+                if publish_job_report(job, args, config.Payload.jobreport):
+                    log.info('wrote job report file')
+                    return True
+                else:
+                    log.info('Failed to write job report file')
```
Comment (PalNilsson) [inline pilot/control/job.py]: change to logger.warning

File: pilot/control/job.py [hunk 9]
```diff
@@ -233,21 +233,38 @@ def send_state(job, args, state, xml=None, metadata=None):
 
     # write the heartbeat message to file if the server is not to be updated by the pilot (Nordugrid mode)
     if not args.update_server:
+        log.info('is_harvester_mode(args) : {0}'.format(is_harvester_mode(args)))
```
Comment (PalNilsson) [inline pilot/control/job.py]: change to logger.debug

File: pilot/control/job.py [hunk 10]
```diff
@@ -233,21 +233,38 @@ def send_state(job, args, state, xml=None, metadata=None):
...
-                publish_job_report(job, args, config.Payload.jobreport)
-                return True
+        if is_harvester_mode(args):
+            # write part of the heartbeat message to worker attributes files needed by Harvester
+            path = get_worker_attributes_file(args)
+            # add jobStatus (state) for Harvester
+            data['jobStatus'] = state
+            # publish work report
+            if publish_work_report(data, path):
+                log.info('wrote to workerAttributesFile %s' % path)
```
Comment (PalNilsson) [inline pilot/control/job.py]: change to logger.debug

File: pilot/util/harvester.py [hunk 11]
```diff
@@ -46,6 +51,7 @@ def get_job_request_file_name():
     :return: job request file name.
     """
 
+    #logger.debug('config.Harvester.__dict__ : {0}'.format(config.Harvester.__dict__))
     return join(environ['PILOT_HOME'], config.Harvester.job_request_file)
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: did you mean os.path.join?

File: pilot/util/harvester.py [hunk 12]
```diff
@@ -126,11 +132,14 @@ def get_event_status_file(args):
...
+
+    logger.debug('config.Harvester.__dict__ : {0}'.format(config.Harvester.__dict__))
+
     if args.harvester_workdir != '':
         work_dir = args.harvester_workdir
     else:
         work_dir = environ['PILOT_HOME']
-    event_status_file = config.Harvester.StageOutnFile
+    event_status_file = config.Harvester.stageoutnfile
     event_status_file = join(work_dir, event_status_file)
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: did you mean os.path.join?

File: pilot/util/harvester.py [hunk 13]
```diff
@@ -145,11 +154,14 @@ def get_worker_attributes_file(args):
...
+
+    logger.debug('config.Harvester.__dict__ : {0}'.format(config.Harvester.__dict__))
+
     if args.harvester_workdir != '':
         work_dir = args.harvester_workdir
     else:
         work_dir = environ['PILOT_HOME']
-    worker_attributes_file = config.Harvester.workerAttributesFile
+    worker_attributes_file = config.Harvester.workerattributesfile
     worker_attributes_file = join(work_dir, worker_attributes_file)
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: did you mean os.path.join?

File: pilot/util/harvester.py [hunk 14]
```diff
@@ -267,6 +294,7 @@ def publish_job_report(job, args, job_report_file="jobReport.json"):
     :param args: Pilot arguments object.
     :param job_report_file: name of job report (string).
     :raises FileHandlingFailure: in case of IOError.
+    :return True or False
     """
 
     src_file = join(job.workdir, job_report_file)
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: did you mean os.path.join?

**Comment by PalNilsson** [comment]
Closing for now since it is blocking a pilot release. To be reopened

**Comment by PalNilsson** [comment]
For some reason git doesn't let me reopen. Doug you may want to recreate the PR. But check the comments above first. I will be back next Tuesday and can try again then, otherwise.

**Comment by PalNilsson** [comment]
Now it reopened for some reason..

File: pilot/util/harvester.py (hunk 1 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: It is used for occasional debugging

File: pilot/util/harvester.py (hunk 5 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: fixed

File: pilot/control/job.py (hunk 10 above)
Comment (dougbenjamin) [reply pilot/control/job.py]: changed

File: pilot/util/harvester.py (hunk 5 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: fixed in now commit

File: pilot/util/harvester.py (hunk 6 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: done

File: pilot/util/harvester.py (hunk 4 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: done

File: pilot/util/harvester.py (hunk 14 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: yes it works with the import but will change for readability

File: pilot/util/harvester.py (hunk 11 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: yes change to os.path.join for readability

File: pilot/control/job.py (hunk 9 above)
Comment (dougbenjamin) [reply pilot/control/job.py]: changed

File: pilot/control/job.py (hunk 7 above)
Comment (dougbenjamin) [reply pilot/control/job.py]: changed to debug

File: pilot/control/job.py (hunk 8 above)
Comment (dougbenjamin) [reply pilot/control/job.py]: changed to debug

File: pilot/util/harvester.py (hunk 13 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: yes changed to os.path.join for readability

**Comment by dougbenjamin** [comment]
made most changes in comments

File: pilot/util/harvester.py [hunk 15]
```diff
@@ -244,19 +259,34 @@ def publish_work_report(work_report=None, worker_attributes_file="worker_attribu
...
+                del(work_report["outputfiles"])
+            if "inputfiles" in work_report:
+                del (work_report["inputfiles"])
+            if "xml" in work_report:
+                del (work_report["xml"])
+            if write_json(worker_attributes_file, work_report):
+                logger.info("work report published: {0}".format(work_report))
+                return True
+            else:
+                logger.info("work report publish failed: {0}".format(work_report))
```
Comment (PalNilsson) [inline pilot/util/harvester.py]: Use logger.warning here

Comment (PalNilsson) [reply pilot/util/harvester.py]: (I'm ignorant to the difference between logger.warning/error btw. if it's a fatal error i think i normally use logger.fatal)

**Comment by PalNilsson** [comment]
BTW add "  # noqa: C901" after the "def send_state(job, args, state, xml=None, metadata=None):" in job.py to ignore the "too complex" flake8 error. This can be sorted out later.

File: pilot/util/harvester.py (hunk 15 above)
Comment (dougbenjamin) [reply pilot/util/harvester.py]: done changed else from info to error

**Pull request merged**